composer/algorithms/agc/agc.py

```diff
@@ -0,0 +1,161 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+"""Core automatic gradient clipping classes and functions."""
+
+from __future__ import annotations
+
+import logging
+from typing import Optional
+
+import torch
+
+from composer.core import Algorithm, Event, State
+from composer.loggers import Logger
+
+log = logging.getLogger(__name__)
+
+__all__ = ["AGC", "apply_agc"]
+
+
+def apply_agc(
+    model: torch.nn.Module,
+    clipping_threshold: float = 0.01,
+    eps: float = 1e-3,
+) -> None:
+    """Clips all gradients in model based on ratio of gradient norms to parameter norms.
+
+    Example:
+         .. testcode::
+
+            from composer.algorithms.agc import apply_agc
+            apply_agc(model=model)
+
+
+    Args:
+        model (torch.nn.Module): The model being trained.
+        clipping_threshold (float, optional): The largest acceptable ratio between grad
+            norms and parameter norms before clipping is done.
+        eps (float, optional): Minimum value that weight norms are clamped to.
+    """
+    for param in model.parameters():
+        if param.grad is None:
+            continue
+
+        # Detach weights and gradients, so the clipping operation is not added to
+        # computational graph.
+        weights = param.detach()
+        grad = param.grad.detach()
+
+        # Get clipped version of gradients.
+        clipped_grad = _get_clipped_gradients(weights, grad)
+
+        # Copy clipped gradients into param.grad attribute, so they can be accessed by
+        # optimizer.
+        grad.copy_(clipped_grad)
```

Comment:
This will increase the gradient memory by 2x, I think we need to do gradient scaling in-place, simlar to how `torch.nn.utils.clip_grad` does it: https://pytorch.org/docs/stable/_modules/torch/nn/utils/clip_grad.html#clip_grad_norm_

I also double-checked whether we can in-place multiply a tensor (grads) by a vector (unit-wise scales), and I think it works:
```
>>> a
tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])
>>> b
tensor([[3.],
        [4.],
        [5.]])
>>> a.mul_(b)
tensor([[3., 3., 3., 3.],
        [4., 4., 4., 4.],
        [5., 5., 5., 5.]])
```

Reply:
Thanks, @abhi-mosaic! Good call! Should be fixed in 16148f1. PTAL. Thanks!